### DISCUSSION TITLE
Question about 2.11 beta

### DISCUSSION THREAD

**Discussion by candlerb** [netbox-community/netbox#6094] [General]
> Parent Interface Assignments (#1519)
>
> Virtual interfaces can now be assigned to a "parent" physical interface by setting the parent field on the interface object. This is helpful for associating subinterfaces with their physical counterpart. For example, you might assign virtual interfaces Gi0/0.100 and Gi0/0.200 as children of the physical interface Gi0/0.

Just checking: does the parent have to be a *physical* interface?

Example 1: Virtual machine with eth0 and eth0.100.  I would like to configure eth0.100 to have eth0 as its parent.
Example 2: Device has a LAG, and a virtual interface vlan256.  I would like to configure vlan256 to have LAG 1 as its parent

**Comment by candlerb** [comment]
I can answer this now from beta.netbox.dev, and the answer is that parent interfaces can only be physical.

I can see why it might be problematic to allow virtual interfaces as parent (particularly avoiding cycles), but here's an example where I'd find it useful.

Consider pfSense as a "firewall-on-a-stick".  If I run it on a physical device, where the external interface is "ix0", this feature will let me associate the "WAN" and "LAN" interfaces with ix0.  (Today, I just create WAN and LAN with no explicit association).

However if I run pfSense as a VM, where the external interface is "vmnet0", and also create "WAN" and "LAN", I won't be able to associate them with the external interface.

**Reply by jeremystretch** [reply]
> and the answer is that parent interfaces can only be physical.

You can designate virtual parent interfaces for virtual interfaces. Currently we only support device interfaces though; I don't think an argument was ever made for VM interface parents.

**Reply by candlerb** [reply]
> You can designate virtual parent interfaces for virtual interfaces. 

That doesn't seem to be true.  Example at https://beta.netbox.dev/dcim/devices/98/

I have created virtual interfaces vlan123 and vlan123456.  I have set a physical interface as the parent of vlan123, but vlan123456 doesn't allow me to set a virtual interface as its parent: vlan123 is not available in the [drop-down list](https://beta.netbox.dev/dcim/interfaces/1166/edit/).

![image](https://user-images.githubusercontent.com/44789/113755810-a7761b80-9708-11eb-905e-8cb17d5497e8.png)

**Reply by tyler-8** [reply]
Parent interfaces on VMs would be useful for most of the same reasons they were introduced for devices. The CSR1000v, for example, can have sub-interfaces (GigabitEthernet1.100) just like a physical Cisco router. That sub-interface is treated as a "child" interface no differently.

**Reply by jeremystretch** [reply]
@candlerb I'm not sure what's going on there, and the demo data has been cleared out. However, I was able to create a virtual `irb` interface and assign to it three virtual child interfaces without issue:

![Screenshot_2021-04-07 dmi01-binghamton-rtr01 - NetBox](https://user-images.githubusercontent.com/13487278/113924824-4443c200-97b8-11eb-8eb9-58b1ac73c180.png)

**Comment by candlerb** [comment]
Hmm, weird, but it's easy to reproduce: here's the exact sequence of steps:

* Devices > Add
    * Role: Access switch
    * Manufacturer: Cisco
    * Device type: C9200-48P
    * Site: Butler Communications
 * Add Interface (blue button at bottom)
     * Name: vlan123
     * Create and add more
     * Name: vlan123456
     * Create
* Change to "100 per page" to see them
* Click edit (pencil) next to vlan123
    * Select parent interface GigabitEthernet1/0/1
    * Update
* Click edit (pencil) next to vlan123456
    * Scroll through parent interface list
    * No vlan interfaces are visible as parent

Until the instance is wiped, this is visible at https://beta.netbox.dev/dcim/interfaces/1166/edit/?return_url=/dcim/devices/98/interfaces/

**Comment by candlerb** [comment]
Incidentally, the name prefix "vlan" isn't significant.  I created virtual interfaces "asdf" and "zxcv" as well, and if I edit zxcv (say), then none of the other virtual interfaces are available to select as parent.

![image](https://user-images.githubusercontent.com/44789/113928622-f6927e00-97e6-11eb-9e59-edfe5593c1c2.png)

**Comment by candlerb** [comment]
This is *really* weird, it may be a caching issue.  I came back, added interface "irb", Create and add more, and then "irb.100" - and then the full set of virtual interfaces was available to select from, including asdf / zxcv / vlan123 / vlan123456.

But then when I go back to vlan123456, and try to select a parent, it's still not showing any virtual interfaces as parent options.

**Reply by candlerb** [reply]
Actually, if I go back to `irb.100`, and try to *change* its parent interface, it's not showing any of the virtual interfaces as options any more.

OK, so what it appears to be is:
* _Creating_ a new virtual interface lets you select a virtual parent: https://beta.netbox.dev/dcim/interfaces/add/?device=98&return_url=/dcim/devices/98/interfaces/
* _Editing_ a virtual interface doesn't let you select a virtual parent: https://beta.netbox.dev/dcim/interfaces/1166/edit/?return_url=/dcim/devices/98/interfaces/

**Reply by jeremystretch** [reply]
Ah, I got it. We use two different forms for interfaces: InterfaceForm and InterfaceCreateForm. The former was filtering by kind but the later wasn't. When I tried to replicate it earlier I was assigning the parent during creation, so I didn't hit the bug. :slightly_smiling_face: 

Anyway, it's an easy fix thankfully. Thanks for tracking this down!

**Reply by jeremystretch** [reply]
Fixed this under #6106 FYI
